go-controller/pkg/cluster/master.go

```diff
@@ -373,21 +393,103 @@ func (cluster *OvnClusterController) addNode(node *kapi.Node) error {
+		logrus.Infof("Allocated node %s HostSubnet %s", node.Name, sn.String())
+		return newHostSubnet(sn.String(), possibleSubnet)
 	}
-	return fmt.Errorf("error allocating netork for node %s: No more allocatable ranges", node.Name)
+	return nil, fmt.Errorf("error allocating network for node %s: No more allocatable ranges", node.Name)
 }
 
-func (cluster *OvnClusterController) deleteNode(node *kapi.Node) error {
+func (cluster *OvnClusterController) ensureNodeLogicalNetwork(node *kapi.Node, hsn *hostSubnet) error {
+	nodeLRPMac, stderr, err := util.RunOVNNbctl("--if-exist", "get", "logical_router_port", "rtos-"+node.Name, "mac")
+	if err != nil {
+		logrus.Errorf("Failed to get logical router port,stderr: %q, error: %v", stderr, err)
+		return err
+	}
+	if nodeLRPMac == "" {
+		nodeLRPMac = util.GenerateMac()
+	}
+
+	// Create a router port and provide it the first address on the node's host subnet
+	_, stderr, err = util.RunOVNNbctl("--may-exist", "lrp-add", OvnClusterRouter, "rtos-"+node.Name, nodeLRPMac, hsn.firstIP)
+	if err != nil {
+		logrus.Errorf("Failed to add logical port to router, stderr: %q, error: %v", stderr, err)
+		return err
+	}
+
+	// Create a logical switch and set its subnet.
+	stdout, stderr, err := util.RunOVNNbctl("--", "--may-exist", "ls-add", node.Name,
+		"--", "set", "logical_switch", node.Name, "other-config:subnet="+hsn.subnet.String(), "external-ids:gateway_ip="+hsn.firstIP)
+	if err != nil {
+		logrus.Errorf("Failed to create a logical switch %v, stdout: %q, stderr: %q, error: %v", node.Name, stdout, stderr, err)
+		return err
+	}
+
+	// Connect the switch to the router.
+	stdout, stderr, err = util.RunOVNNbctl("--", "--may-exist", "lsp-add", node.Name, "stor-"+node.Name,
+		"--", "set", "logical_switch_port", "stor-"+node.Name, "type=router", "options:router-port=rtos-"+node.Name, "addresses="+"\""+nodeLRPMac+"\"")
+	if err != nil {
+		logrus.Errorf("Failed to add logical port to switch, stdout: %q, stderr: %q, error: %v", stdout, stderr, err)
+		return err
+	}
+
+	// Add our cluster TCP and UDP load balancers to the node switch
+	if cluster.TCPLoadBalancerUUID == "" {
+		return fmt.Errorf("TCP cluster load balancer not created")
+	}
+	stdout, stderr, err = util.RunOVNNbctl("set", "logical_switch", node.Name, "load_balancer="+cluster.TCPLoadBalancerUUID)
+	if err != nil {
+		logrus.Errorf("Failed to set logical switch %v's loadbalancer, stdout: %q, stderr: %q, error: %v", node.Name, stdout, stderr, err)
+		return err
+	}
+
+	if cluster.UDPLoadBalancerUUID == "" {
+		return fmt.Errorf("UDP cluster load balancer not created")
+	}
+	stdout, stderr, err = util.RunOVNNbctl("add", "logical_switch", node.Name, "load_balancer", cluster.UDPLoadBalancerUUID)
+	if err != nil {
+		logrus.Errorf("Failed to add logical switch %v's loadbalancer, stdout: %q, stderr: %q, error: %v", node.Name, stdout, stderr, err)
+		return err
+	}
+
+	return nil
+}
+
+func (cluster *OvnClusterController) addNode(node *kapi.Node) (err error) {
```

Comment:
I forget what happens when the informers are set up - do we get an `Add` or an `Update` event for every existing node? If it's `Update`, then we'll have to deal with that.

Reply:
@squeed we get an add for every existing object; see pkg/factory/factory.go when a handler (eg watch) gets added. Regular kube informers do this, so we emulate it in our shared informer.

```
	// Send existing items to the handler's add function; informers usually
	// do this but since we share informers, it's long-since happened so
	// we must emulate that here
	for _, obj := range existingItems {
		inf.handlers[handlerID].OnAdd(obj)
	}
```